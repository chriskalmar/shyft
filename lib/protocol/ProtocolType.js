
import {
  passOrThrow,
  isFunction,
} from '../util';

import { isDataType } from '../datatype/DataType';


class ProtocolType {

  constructor (setup = {}) {

    const {
      name,
      description,
      isProtocolDataType,
    } = setup

    passOrThrow(name, () => 'Missing protocol type name')
    passOrThrow(description, () => `Missing description for protocol type '${name}'`)

    passOrThrow(
      isFunction(isProtocolDataType),
      () => `Protocol type '${name}' needs to implement isProtocolDataType()`
    )

    this.name = name
    this.description = description
    this.isProtocolDataType = isProtocolDataType

    this._dataTypeMap = {}
  }


  addDataTypeMap(schemaDataType, protocolDataType) {

    passOrThrow(
      isDataType(schemaDataType),
      () => `Provided schemaDataType is not a valid data type in '${this.name}'`
    )

    passOrThrow(
      this.isProtocolDataType(protocolDataType),
      () => `Provided protocolDataType for '${schemaDataType}' is not a valid protocol data type in '${this.name}'`
    )

    passOrThrow(
      !this._dataTypeMap[ schemaDataType.name ],
      () => `Data type mapping for '${schemaDataType.name}' already registered with protocol type '${this.name}'`
    )

    this._dataTypeMap[ schemaDataType.name ] = protocolDataType
  }


  convertToProtocolDataType (schemaDataType) {

    passOrThrow(
      isDataType(schemaDataType),
      () => `Provided schemaDataType is not a valid data type in '${this.name}'`
    )

    passOrThrow(
      this._dataTypeMap[ schemaDataType.name ],
      () => `No data type mapping found for '${schemaDataType.name}' in protocol type '${this.name}'`
    )

    return this._dataTypeMap[ schemaDataType.name ]
  }


  toString() {
    return this.name
  }

}


export default ProtocolType


export const isProtocolType = (obj) => {
  return (obj instanceof ProtocolType)
}
