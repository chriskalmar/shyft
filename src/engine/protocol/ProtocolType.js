import { passOrThrow, isFunction } from '../util';

import { isDataType } from '../datatype/DataType';
import { isProtocolConfiguration } from './ProtocolConfiguration';

export class ProtocolType {
  constructor(setup = {}) {
    const { name, description, isProtocolDataType } = setup;

    passOrThrow(name, () => 'Missing protocol type name');
    passOrThrow(
      description,
      () => `Missing description for protocol type '${name}'`,
    );

    passOrThrow(
      isFunction(isProtocolDataType),
      () => `Protocol type '${name}' needs to implement isProtocolDataType()`,
    );

    this.name = name;
    this.description = description;
    this.isProtocolDataType = isProtocolDataType;

    this._dataTypeMap = {};
    this._dynamicDataTypeMap = [];
  }

  addDataTypeMap(schemaDataType, protocolDataType) {
    passOrThrow(
      isDataType(schemaDataType),
      () =>
        `Provided schemaDataType is not a valid data type in '${this.name}'`,
    );

    passOrThrow(
      this.isProtocolDataType(protocolDataType),
      () =>
        `Provided protocolDataType for '${String(
          schemaDataType,
        )}' is not a valid protocol data type in '${this.name}'`,
    );

    passOrThrow(
      !this._dataTypeMap[schemaDataType.name],
      () =>
        `Data type mapping for '${schemaDataType.name}' already registered with protocol type '${this.name}'`,
    );

    this._dataTypeMap[schemaDataType.name] = protocolDataType;
  }

  addDynamicDataTypeMap(schemaDataTypeDetector, protocolDataType) {
    passOrThrow(
      isFunction(schemaDataTypeDetector),
      () =>
        `Provided schemaDataTypeDetector is not a valid function in '${this.name}', ` +
        `got this instead: ${String(schemaDataTypeDetector)}`,
    );

    passOrThrow(
      this.isProtocolDataType(protocolDataType) || isFunction(protocolDataType),
      () =>
        `Provided protocolDataType for '${String(
          schemaDataTypeDetector,
        )}' is not a valid protocol data type or function in '${this.name}', ` +
        `got this instead: ${String(protocolDataType)}`,
    );

    this._dynamicDataTypeMap.push({
      schemaDataTypeDetector,
      protocolDataType,
    });
  }

  convertToProtocolDataType(schemaDataType, sourceName, asInput) {
    const foundDynamicDataType = this._dynamicDataTypeMap.find(
      ({ schemaDataTypeDetector }) => schemaDataTypeDetector(schemaDataType),
    );
    if (foundDynamicDataType) {
      const protocolDataType = foundDynamicDataType.protocolDataType;

      if (isFunction(protocolDataType)) {
        const attributeType = schemaDataType;
        return protocolDataType(attributeType, sourceName, asInput);
      }

      return protocolDataType;
    }

    passOrThrow(
      isDataType(schemaDataType),
      () =>
        `Provided schemaDataType is not a valid data type in protocol type '${this.name}'`,
    );

    passOrThrow(
      this._dataTypeMap[schemaDataType.name],
      () =>
        `No data type mapping found for '${schemaDataType.name}' in protocol type '${this.name}'`,
    );

    return this._dataTypeMap[schemaDataType.name];
  }

  setProtocolConfiguration(protocolConfiguration) {
    passOrThrow(
      isProtocolConfiguration(protocolConfiguration),
      () => 'ProtocolType expects a valid protocolConfiguration',
    );

    this.protocolConfiguration = protocolConfiguration;
  }

  getProtocolConfiguration() {
    passOrThrow(
      this.protocolConfiguration,
      () => 'ProtocolType is missing a valid protocolConfiguration',
    );

    return this.protocolConfiguration;
  }

  toString() {
    return this.name;
  }
}

export const isProtocolType = obj => {
  return obj instanceof ProtocolType;
};
