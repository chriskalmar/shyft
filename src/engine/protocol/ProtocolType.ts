/* eslint-disable @typescript-eslint/no-explicit-any */

import { passOrThrow, isFunction } from '../util';

import { DataType, DataTypeFunction, isDataType } from '../datatype/DataType';
import {
  ProtocolConfiguration,
  isProtocolConfiguration,
} from './ProtocolConfiguration';
import { ComplexDataType } from '../..';
import { isComplexDataType } from '../datatype/ComplexDataType';

export type ProtocolTypeSetup = {
  name?: string;
  description?: string;
  // isProtocolDataType?: Function;
  isProtocolDataType?: (protocolDataType: any) => boolean;
};

// export interface ProtocolDataTypeFunction {
//   schemaDataType: DataType;
//   attribute: string;
//   asInput?: boolean;
// }

// how to properly define this type ?
export type ProtocolDataType = {
  name: string | Function;
};

type DataTypeMap = {
  [name: string]: ProtocolDataType;
};

type DynamicDataTypeMap = {
  schemaDataTypeDetector: Function;
  protocolDataType: ProtocolDataType;
};

export class ProtocolType {
  name: string;
  description: string;
  isProtocolDataType: (protocolDataType: any) => boolean;
  protocolConfiguration: ProtocolConfiguration;

  private _dataTypeMap: DataTypeMap;
  private _dynamicDataTypeMap: DynamicDataTypeMap[];

  constructor(setup: ProtocolTypeSetup = {} as ProtocolTypeSetup) {
    const { name, description, isProtocolDataType } = setup;

    passOrThrow(name, () => 'Missing protocol type name');
    passOrThrow(
      description,
      () => `Missing description for protocol type '${name}'`,
    );

    passOrThrow(
      isFunction(isProtocolDataType),
      () => `Protocol type '${name}' needs to implement isProtocolDataType()`,
    );

    this.name = name;
    this.description = description;
    this.isProtocolDataType = isProtocolDataType;

    this._dataTypeMap = {};
    this._dynamicDataTypeMap = [];
  }

  addDataTypeMap(
    schemaDataType: DataType | DataTypeFunction,
    protocolDataType: ProtocolDataType,
  ): void {
    passOrThrow(
      isDataType(schemaDataType),
      () =>
        `Provided schemaDataType is not a valid data type in '${this.name}'`,
    );

    passOrThrow(
      this.isProtocolDataType(protocolDataType),
      () =>
        `Provided protocolDataType for '${String(
          schemaDataType,
        )}' is not a valid protocol data type in '${this.name}'`,
    );

    passOrThrow(
      !this._dataTypeMap[schemaDataType.name],
      () =>
        `Data type mapping for '${schemaDataType.name}' already registered with protocol type '${this.name}'`,
    );

    this._dataTypeMap[schemaDataType.name] = protocolDataType;
  }

  addDynamicDataTypeMap(
    schemaDataTypeDetector: Function,
    protocolDataType: any,
    // protocolDataType: ProtocolType,
  ): void {
    passOrThrow(
      isFunction(schemaDataTypeDetector),
      () =>
        `Provided schemaDataTypeDetector is not a valid function in '${this.name}', ` +
        `got this instead: ${String(schemaDataTypeDetector)}`,
    );

    passOrThrow(
      this.isProtocolDataType(protocolDataType) || isFunction(protocolDataType),
      () =>
        `Provided protocolDataType for '${String(
          schemaDataTypeDetector,
        )}' is not a valid protocol data type or function in '${this.name}', ` +
        `got this instead: ${String(protocolDataType)}`,
    );

    this._dynamicDataTypeMap.push({
      schemaDataTypeDetector,
      protocolDataType,
    });
  }

  convertToProtocolDataType(
    schemaDataType: DataType | ComplexDataType | DataTypeFunction,
    sourceName?: string,
    asInput?: boolean,
  ): any {
    const foundDynamicDataType = this._dynamicDataTypeMap.find(
      ({ schemaDataTypeDetector }) => schemaDataTypeDetector(schemaDataType),
    );
    if (foundDynamicDataType) {
      const protocolDataType = foundDynamicDataType.protocolDataType;

      if (isFunction(protocolDataType)) {
        const attributeType = schemaDataType;
        const protocolDataTypeFn = protocolDataType as Function;
        return protocolDataTypeFn(attributeType, sourceName, asInput);
      }

      return protocolDataType;
    }

    if (isDataType(schemaDataType)) {
      passOrThrow(
        this._dataTypeMap[schemaDataType.name],
        () =>
          `No data type mapping found for '${schemaDataType.name}' in protocol type '${this.name}'`,
      );

      return this._dataTypeMap[schemaDataType.name];
    }

    throw new Error(
      `Provided schemaDataType is not a valid data type in protocol type '${this.name}'`,
    );
  }

  setProtocolConfiguration(protocolConfiguration): void {
    passOrThrow(
      isProtocolConfiguration(protocolConfiguration),
      () => 'ProtocolType expects a valid protocolConfiguration',
    );

    this.protocolConfiguration = protocolConfiguration;
  }

  getProtocolConfiguration(): ProtocolConfiguration {
    passOrThrow(
      this.protocolConfiguration,
      () => 'ProtocolType is missing a valid protocolConfiguration',
    );

    return this.protocolConfiguration;
  }

  toString(): string {
    return this.name;
  }
}

export const isProtocolType = (obj: unknown): obj is ProtocolType => {
  return obj instanceof ProtocolType;
};
