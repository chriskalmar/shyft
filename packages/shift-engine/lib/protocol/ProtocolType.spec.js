
import ProtocolType, { isProtocolType } from './ProtocolType';
import {
  DataTypeID,
  DataTypeString,
  DataTypeInteger,
  DataTypeBoolean,
} from '../datatype/dataTypes';


describe('ProtocolType', () => {

  const ProtocolTypeREST = new ProtocolType({
    name: 'ProtocolTypeREST',
    description: 'REST API protocol',
    isProtocolDataType(protocolDataType) {
      return (typeof protocolDataType  === 'object' && protocolDataType.name)
    }
  })

  const ProtocolDataTypeID = {
    name: 'ProtocolDataTypeID'
  }

  const ProtocolDataTypeString = {
    name: 'ProtocolDataTypeString'
  }

  const ProtocolDataTypeInteger = {
    name: 'ProtocolDataTypeInteger'
  }


  it('should reject invalid protocol type definitions', () => {

    function fn1() {
      new ProtocolType({}) // eslint-disable-line no-new
    }

    function fn2() {
      new ProtocolType({ // eslint-disable-line no-new
        name: 'REST',
      })
    }

    expect(fn1).toThrow();
    expect(fn2).toThrow();

  })


  it('should implement isProtocolDataType', () => {

    function fn() {
      new ProtocolType({ // eslint-disable-line no-new
        name: 'REST',
        description: 'REST protocol type'
      })
    }

    expect(fn).toThrow();

  })


  it('should recognize object of type ProtocolType', () => {

    const result = isProtocolType(ProtocolTypeREST)
    expect(result).toBe(true)
  })


  it('should recognize non-ProtocolType objects', () => {

    const result1 = isProtocolType({})
    const result2 = isProtocolType(123)
    const result3 = isProtocolType('test')

    expect(result1).toBe(false)
    expect(result2).toBe(false)
    expect(result3).toBe(false)
  })



  it('should map data types to protocol data types', () => {

    ProtocolTypeREST.addDataTypeMap(DataTypeID, ProtocolDataTypeID)
    ProtocolTypeREST.addDataTypeMap(DataTypeString, ProtocolDataTypeString)

  })


  it('should reject invalid data type mappings', () => {

    function fn1() {
      ProtocolTypeREST.addDataTypeMap('something')
    }

    function fn2() {
      ProtocolTypeREST.addDataTypeMap(DataTypeID)
    }

    function fn3() {
      ProtocolTypeREST.addDataTypeMap(DataTypeString, {})
    }

    expect(fn1).toThrow();
    expect(fn2).toThrow();
    expect(fn3).toThrow();

  })


  it('should reject duplicate data type mappings', () => {

    ProtocolTypeREST.addDataTypeMap(DataTypeInteger, ProtocolDataTypeInteger)

    function fn() {
      ProtocolTypeREST.addDataTypeMap(DataTypeInteger, ProtocolDataTypeInteger)
    }

    expect(fn).toThrow();
  })



  it('should convert to protocol data types', () => {

    const result1 = ProtocolTypeREST.convertToProtocolDataType(DataTypeID)
    const result2 = ProtocolTypeREST.convertToProtocolDataType(DataTypeString)
    const result3 = ProtocolTypeREST.convertToProtocolDataType(DataTypeInteger)

    expect(result1).toEqual(ProtocolDataTypeID)
    expect(result2).toEqual(ProtocolDataTypeString)
    expect(result3).toEqual(ProtocolDataTypeInteger)

  })


  it('should throw if unknown data type mapping is requested', () => {

    function fn1() {
      ProtocolTypeREST.convertToProtocolDataType('DataTypeInteger')
    }

    function fn2() {
      ProtocolTypeREST.convertToProtocolDataType(DataTypeBoolean)
    }

    expect(fn1).toThrow();
    expect(fn2).toThrow();

  })


})
