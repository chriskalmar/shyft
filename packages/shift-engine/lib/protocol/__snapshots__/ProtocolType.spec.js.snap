// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtocolType should implement isProtocolDataType 1`] = `"Protocol type 'REST' needs to implement isProtocolDataType()"`;

exports[`ProtocolType should reject duplicate data type mappings 1`] = `"Data type mapping for 'DataTypeInteger' already registered with protocol type 'ProtocolTypeREST'"`;

exports[`ProtocolType should reject invalid data type mappings 1`] = `"Provided schemaDataType is not a valid data type in 'ProtocolTypeREST'"`;

exports[`ProtocolType should reject invalid data type mappings 2`] = `"Provided protocolDataType for 'DataTypeID' is not a valid protocol data type in 'ProtocolTypeREST'"`;

exports[`ProtocolType should reject invalid data type mappings 3`] = `"Provided protocolDataType for 'DataTypeString' is not a valid protocol data type in 'ProtocolTypeREST'"`;

exports[`ProtocolType should reject invalid protocol type definitions 1`] = `"Missing protocol type name"`;

exports[`ProtocolType should reject invalid protocol type definitions 2`] = `"Missing description for protocol type 'REST'"`;

exports[`ProtocolType should throw if unknown data type mapping is requested 1`] = `"Provided schemaDataType is not a valid data type in protocol type 'ProtocolTypeREST'"`;

exports[`ProtocolType should throw if unknown data type mapping is requested 2`] = `"No data type mapping found for 'DataTypeBoolean' in protocol type 'ProtocolTypeREST'"`;
